File: examples/nodejs/custom_tool.ts
/** Example: Register a custom tool with the agent. */

import {
  Agent, resolveConfig, ITool, ToolContext, successResult, errorResult,
} from '../src/index.js';

class CalculatorTool implements ITool {
  readonly name = 'calculator';
  readonly description = 'Perform basic arithmetic: add, subtract, multiply, divide.';
  readonly isReadOnly = true;

  readonly inputSchema = {
    type: 'object',
    properties: {
      operation: { type: 'string', enum: ['add', 'subtract', 'multiply', 'divide'] },
      a: { type: 'number' },
      b: { type: 'number' },
    },
    required: ['operation', 'a', 'b'],
  };

  async call(input: Record<string, unknown>, _ctx: ToolContext) {
    const operation = input.operation as string;
    const a = Number(input.a);
    const b = Number(input.b);

    if (operation === 'divide' && b === 0) {
      return errorResult('Cannot divide by zero');
    }

    let result = 0;
    switch (operation) {
      case 'add': result = a + b; break;
      case 'subtract': result = a - b; break;
      case 'multiply': result = a * b; break;
      case 'divide': result = a / b; break;
      default: return errorResult(`Unknown operation: ${operation}`);
    }
    return successResult(String(result));
  }
}

async function main() {
  const config = resolveConfig({ apiKey: process.env.ANTHROPIC_AUTH_TOKEN || '' });
  const agent = new Agent(config);
  agent.registerTool(new CalculatorTool());

  for await (const event of agent.run('What is 123 multiplied by 456?')) {
    if (event.type === 'message_delta') {
      process.stdout.write(event.data.text as string);
    } else if (event.type === 'complete') {
      console.log('\n\n[Complete]', event.data.final_content);
    } else if (event.type === 'error') {
      console.error('\n[Error]', event.data.message);
    }
  }
}

main().catch(console.error);

File: examples/rust/custom_tool.rs
// Example: Register a custom tool with the agent.
use agentlib::{Agent, AgentConfig, Tool, ToolContext, ToolResult};
use async_trait::async_trait;
use serde_json::json;
use std::env;

struct CalculatorTool;

#[async_trait]
impl Tool for CalculatorTool {
    fn name(&self) -> &str { "calculator" }
    fn description(&self) -> &str { "Perform basic arithmetic." }
    fn is_read_only(&self) -> bool { true }

    fn input_schema(&self) -> serde_json::Value {
        json!({
            "type": "object",
            "properties": {
                "operation": { "type": "string", "enum": ["add", "subtract", "multiply", "divide"] },
                "a": { "type": "number" },
                "b": { "type": "number" }
            },
            "required": ["operation", "a", "b"]
        })
    }

    async fn call(&self, input: serde_json::Value, _ctx: &ToolContext) -> Result<ToolResult, agentlib::ToolError> {
        let op = input["operation"].as_str().unwrap_or("");
        let a = input["a"].as_f64().unwrap_or(0.0);
        let b = input["b"].as_f64().unwrap_or(0.0);

        let result = match op {
            "add" => a + b,
            "subtract" => a - b,
            "multiply" => a * b,
            "divide" => {
                if b == 0.0 {
                    return Ok(ToolResult::error("Cannot divide by zero"));
                }
                a / b
            }
            _ => return Ok(ToolResult::error(&format!("Unknown operation: {}", op))),
        };

        Ok(ToolResult::success(&result.to_string()))
    }
}

#[tokio::main]
async fn main() {
    let config = AgentConfig {
        api_key: env::var("ANTHROPIC_AUTH_TOKEN").unwrap_or_default(),
        ..Default::default()
    };

    let mut agent = Agent::new(config);
    agent.register_tool(Box::new(CalculatorTool));

    let mut rx = agent.run("What is 123 multiplied by 456?").await.unwrap();
    while let Some(event) = rx.recv().await {
        match event {
            agentlib::Event::MessageDelta { text } => print!("{}", text),
            agentlib::Event::Complete { final_content } => println!("\n\n[Complete] {}", final_content),
            agentlib::Event::Error { message } => println!("\n[Error] {}", message),
            _ => {}
        }
    }
}
